pkg/resource/managed_reconciler.go

```diff
@@ -0,0 +1,438 @@
+// runtime.Scheme. The returned ManagedReconciler reconciles with a dummy, no-op
+// 'external system' by default; callers should supply an ExternalConnector that
+// returns an ExternalClient capable of managing resources in a real system.
+func NewManagedReconciler(m manager.Manager, of ManagedKind, o ...ManagedReconcilerOption) *ManagedReconciler {
+	nm := func() Managed { return MustCreateObject(schema.GroupVersionKind(of), m.GetScheme()).(Managed) }
+
+	// Panic early if we've been asked to reconcile a resource kind that has not
+	// been registered with our controller manager's scheme.
+	_ = nm()
+
+	r := &ManagedReconciler{
+		client:     m.GetClient(),
+		newManaged: nm,
+		shortWait:  defaultManagedShortWait,
+		longWait:   defaultManagedLongWait,
+		managed:    defaultMRManaged(m),
+		external:   defaultMRExternal(),
+	}
+
+	for _, ro := range o {
+		ro(r)
+	}
+
+	return r
+}
+
+// Reconcile a managed resource with an external resource.
+func (r *ManagedReconciler) Reconcile(req reconcile.Request) (reconcile.Result, error) { // nolint:gocyclo
+	// NOTE(negz): This method is a little over our cyclomatic complexity goal.
+	// Be wary of adding additional complexity.
+
+	log.V(logging.Debug).Info("Reconciling", "controller", managedControllerName, "request", req)
+
+	ctx, cancel := context.WithTimeout(context.Background(), managedReconcileTimeout)
+	defer cancel()
+
+	managed := r.newManaged()
+	if err := r.client.Get(ctx, req.NamespacedName, managed); err != nil {
+		// There's no need to requeue if we no longer exist. Otherwise we'll be
+		// requeued implicitly because we return an error.
+		return reconcile.Result{}, errors.Wrap(IgnoreNotFound(err), errGetManaged)
+	}
+
+	external, err := r.external.Connect(ctx, managed)
+	if err != nil {
+		// We'll usually hit this case if our Provider or its secret are missing
+		// or invalid. If this is first time we encounter this issue we'll be
+		// requeued implicitly when we update our status with the new error
+		// condition. If not, we want to try again after a short wait.
+		managed.SetConditions(v1alpha1.ReconcileError(err))
+		return reconcile.Result{RequeueAfter: r.shortWait}, errors.Wrap(r.client.Status().Update(ctx, managed), errUpdateManagedStatus)
+	}
+
+	o, err := external.Observe(ctx, managed)
+	if err != nil {
+		// We'll usually hit this case if our Provider credentials are invalid
+		// or insufficient for observing the external resource type we're
+		// concerned with. If this is the first time we encounter this issue
+		// we'll be requeued implicitly when we update our status with the new
+		// error condition. If not, we want to try again after a short wait.
+		managed.SetConditions(v1alpha1.ReconcileError(err))
+		return reconcile.Result{RequeueAfter: r.shortWait}, errors.Wrap(r.client.Status().Update(ctx, managed), errUpdateManagedStatus)
+	}
+
+	if err := r.managed.Publish(ctx, managed, o.ConnectionDetails); err != nil {
```

Comment:
I got excited there for a second, but I think we do need this case. It's publishing connection details found during the observation of a resource, which may be different from connection details observed during the update of a resource.

If we removed this block we'd have to ensure the `Update` call returned all connection details, even if they were unrelated to the `Update` operation; i.e. `Update` would be responsible for both observing and updating the managed resource. This is actually how some of our controllers were previously designed; the responsibilities of this Reconciler's `Observe` and `Update` methods were shared by a single `Sync` method. I factored them out into two distinct methods for this design because:

* The Sync methods were becoming large and complicated in their responsibilities.
* Observing the external resource early on during the reconcile allows us to determine whether we should create it based on whether it actually exists, rather than because some artifact of the managed resource status thinks it exists.

Reply:
I see, makes sense. There are ways to get rid of that call, like defining a `ConnectionDetails` in Reconciler and override in each step, then `Publish` it in the end. But as you said that'd complicate the `Reconcile` in the sense that it'd have a custom logic for a lower level component(connection) compared to other operations it does.

One option to consider is that we can have only `ExternalObservation` struct and have all `Create/Update/Observe/Delete` return that type, then override the one we'd define in the beginning of `Reconcile` with each call and publish it in the end. Again, complicates the `Reconcile` but at least not toward only `ConnectionDetails` but for any kind of result that we encapsulate with `ExternalObservation`. I don't have strong arguments for that though, simplification maybe?

Reply:
I like the idea of everything returning the same `ExternalObservation` struct, and would be happy to see whether that allows us to reduce the number of `Publish` calls we need.

Reply:
I have tried a few different ways but in each of them the following lines haunted me:
```
// NOTE(negz): This method is a little over our cyclomatic complexity goal.
// Be wary of adding additional complexity.
```

There are ways to make only 1 Publish call but at this point, I think it's probably not worth the added cyclomatic complexity. Using `ExternalObservation` for all return types imposes a requirement on Create/Update/Delete such that they have to say whether `ResourceExists` but there is not really a good way to extract this information. For `Observe`, it's to check `IsNotFound` for the main HTTP request. But for other functions, there could be more than one call and they might return 404 for different reasons. You can see a half-done not-working version of it here https://github.com/muvaf/crossplane/commit/d74c1daeaf357327561b61432e77c1443323d162

Reply:
Your approach in muvaf@d74c1da looks similar to some patterns I played around with, and I also came back to this feeling cleaner in the current implementation. I prefer more `Publish` calls and more readable code than fewer `Publish` calls and more complicated code. Let's leave this as is for the time being; we can always refactor in a future PR if a cleaner pattern comes to us in a dream.